# P2 - Data Analyst

## Importing
Import libraries and CSV. Verify if CSV was imported correctly.

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from bokeh.plotting import figure
from bokeh.charts import Histogram
from bokeh.io import output_notebook, show
from bokeh.layouts import row
import random
from sets import Set

output_notebook()

df = pd.read_csv('titanic_data.csv')

C:\Anaconda2\lib\site-packages\ipykernel\__main__.py:12: DeprecationWarning: the sets module is deprecated


Loading BokehJS ...

In [2]:
df.describe()

C:\Anaconda2\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000         NaN    0.000000   
50%     446.000000    0.000000    3.000000         NaN    0.000000   
75%     668.500000    1.000000    3.000000         NaN    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [3]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

## Data Cleaning
The 'Age' has some NaN and we must fix it.

In [4]:
df.isnull().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

From a total of 891 passengers, we don't have the Age of 177, the Cabin of 687 and the Place of Embark for 2.

We can consider that Cabin is nearly useless and remove this column.

For the Place of Embark and Age, I will remove the passengers from our analysis. I am not happy by doing it, they are over 150 being removed but we will still keep over than 700 passengers to analyse

In [5]:
df_clean = df.copy()

del df_clean['Cabin']

df_clean = df_clean.dropna()

df_clean.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

The column PassengerId and Ticket will not give me any relevant statistical result, so they will be also removed.

In [6]:
del df_clean['PassengerId']
del df_clean['Ticket']

df_clean.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  712.000000  712.000000  712.000000  712.000000  712.000000  712.000000
mean     0.404494    2.240169   29.642093    0.514045    0.432584   34.567251
std      0.491139    0.836854   14.492933    0.930692    0.854181   52.938648
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    1.000000   20.000000    0.000000    0.000000    8.050000
50%      0.000000    2.000000   28.000000    0.000000    0.000000   15.645850
75%      1.000000    3.000000   38.000000    1.000000    1.000000   33.000000
max      1.000000    3.000000   80.000000    5.000000    6.000000  512.329200

## Exploring phase
Now that the data is clean it is time to play with it. Let´s start with Age, we will verify if there is a major difference between the age of those who survived.

In [7]:
print("The total number of passengers were %d, from them only %d survived." 
      %(df_clean['Survived'].count(), df_clean[df_clean['Survived'] == 1]['Survived'].count()))
print("There is %d males and %d females in the cleaned data." 
      %(df_clean[df_clean['Sex'] == 'male']['Survived'].count(),
        df_clean[df_clean['Sex'] == 'female']['Survived'].count()))


The total number of passengers were 712, from them only 288 survived.
There is 453 males and 259 females in the cleaned data.


Let´s check the Age, we will verify if there is a major difference between the age of those who survived.

In [8]:
hist = Histogram(df_clean, values='Age', title="Age of passagers Histogram", plot_width=400, plot_height=400)
hist2 = Histogram(df_clean, values='Age', color='Survived', legend='top_right',
                  title="Age of passagers Histogram if Survived or not", plot_width=400, plot_height=400)

show(row(hist,hist2))

It surprised me that most of the babies (below 5) survived and someone above 75. But it is not much to play with it.

Let's check if there is a difference for 'Sex'.

In [9]:
hist = Histogram(df_clean, values='Survived', color='Sex', legend='top_right',
                  title="Age of passagers Histogram if Survived or not", plot_width=400, plot_height=400)
hist2 = Histogram(df_clean, values='Age', color='Sex', legend='top_right',
                  title="Age of passagers Histogram if Survived or not", plot_width=400, plot_height=400)

show(row(hist,hist2))

Now it really surprised me. The amount of females that survived is a lot bigger them the mount of males.

The age between males and females didn't change a lot. Since we checked something more interesting in the survival, let's play with it.

In [10]:
def survival_rate(sex):
    df_one_sex = df_clean[df_clean['Sex'] == sex]
    count_survivals = df_one_sex[df_one_sex['Survived'] == 1].loc[:,'Survived'].count()
    count_one_sex = df_one_sex['Survived'].count()
    return float(count_survivals)/count_one_sex

survival_rate_male = survival_rate('male')
survival_rate_female = survival_rate('female')

print("The survival rate for males were %.2f and for females were %.2f." %(survival_rate_male, survival_rate_female))

The survival rate for males were 0.21 and for females were 0.75.


I am suprised again, I didn't expect that much difference in the percentages. I will try to statistically indicate that the survival rate for the females in the titanic were higher than the males.

# Statistical analysis
Here will we try to show the statistical significance that the average survival rate for the females in the titanic were higher than the survival rate for the males.

We will draw 1000 random samples of 20 subjects for each, show the histogram of the survival rate for the samples and check if the average survival rate for females is significant higher than for the males.

We will use the critical value of 0.05 and we will use the t-test.

So Hypotesis 0:
average survival rate of females <= average survival rate of males

Alternative Hypotesis:
average survival rate of females > average survival rate of males

In [11]:
SIZE_SAMPLE = 20
TOTAL_SIZE = 1000

def sample(sex,size_sample):
    indexes = Set([])
    while len(indexes) < size_sample:
        df_one_sex = df_clean[df_clean['Sex'] == sex]
        indexes.add(random.choice(df_one_sex.index))
    return indexes

def survival_rate_of_sample(indexes):
    survivals = []
    for index in indexes:
        survivals.append(df_clean.loc[index,'Survived'])
    return float(sum(survivals))/len(survivals)

def survival_rate_list(sex,total_size,size_sample):
    samples_surv_rate = []
    for num_sample in range(total_size):
        samples_surv_rate.append(survival_rate_of_sample(sample(sex,size_sample)))
    return samples_surv_rate
    

survival_rate_list_male = survival_rate_list('male',TOTAL_SIZE,SIZE_SAMPLE)
survival_rate_list_female = survival_rate_list('female',TOTAL_SIZE,SIZE_SAMPLE)

Note: there is a chance of keeping twice the same survival rate (if the 20 samples are the same). Since the chance is extremely small we are ignoring it.

In [12]:
data = pd.DataFrame({'Sex': [], 'survival_rate': []})

for survival_rate_male in survival_rate_list_male:
    data = data.append({'Sex': 'male', 'survival_rate': survival_rate_male},ignore_index=True)
for survival_rate_female in survival_rate_list_female:
    data = data.append({'Sex': 'female', 'survival_rate': survival_rate_female},ignore_index=True)

In [13]:
hist = Histogram(data, values='survival_rate', color='Sex', title="Survival rate for sex Histogram",
                 plot_width=400, plot_height=400, legend='top_center')

show(hist)

We want to check the difference so we will analyse the female survival rate - male survival rate

In [14]:
survival_rate_difference = np.array(survival_rate_list_female) - np.array(survival_rate_list_male)

hist = Histogram(pd.DataFrame({'Survival rate difference': survival_rate_difference}),
                 title="Survival rate difference Histogram", plot_width=400, plot_height=400)

show(hist)

In [15]:
print('The mean is %.2f and the standard deviation is %.2f.' %(survival_rate_difference.mean(),survival_rate_difference.std()))

The mean is 0.55 and the standard deviation is 0.13.


In [16]:
t_statistic = survival_rate_difference.mean()/(survival_rate_difference.std()/SIZE_SAMPLE**0.5)

print('The t-statistic is %.2f.' %t_statistic)

The t-statistic is 19.40.


The degree of freedom is 19.

The t-critical is 1.729. (Remember we are only checking if it is higher, one direction.)

# Conclusion
Since t-statistical is a lot higher than t-critical we can say that we drop the initial hypotesis and assume that the average survival rate for females is significant higher than the average survival rate for males.